packages/react/src/components/Listbox/Listbox.tsx

```diff
@@ -0,0 +1,216 @@
+import React, {
+  forwardRef,
+  useCallback,
+  useState,
+  useLayoutEffect,
+  useEffect
+} from 'react';
+import { ListboxProvider } from './ListboxContext';
+import type { ListboxOption } from './ListboxContext';
+import type { ListboxValue } from './ListboxOption';
+import useSharedRef from '../../utils/useSharedRef';
+
+const keys = ['ArrowUp', 'ArrowDown', 'Home', 'End', 'Enter', ' '];
+
+interface ListboxProps
+  extends Omit<React.HTMLAttributes<HTMLElement>, 'onSelect'> {
+  as?: React.ElementType | string;
+  value?: ListboxValue;
+  navigation?: 'cycle' | 'bound';
+  onSelect?: <T extends HTMLElement = HTMLElement>({
+    target,
+    value
+  }: {
+    target: T;
+    value: ListboxValue;
+  }) => void;
+  onSelectionChange?: ({
+    value
+  }: {
+    previousValue: ListboxValue;
+    value: ListboxValue;
+  }) => void;
+  onActiveChange?: (option: ListboxOption) => void;
+}
+
+// id for listbox options should always be defined since it should
+// be provide via the author, or auto-generated via the component
+const getOptionId = (option: ListboxOption): string =>
+  option.element.getAttribute('id') as string;
+
+const isDisabledOption = (option: ListboxOption): boolean =>
+  option.element.getAttribute('aria-disabled') === 'true';
+
+const optionMatchesValue = (option: ListboxOption, value: unknown): boolean =>
+  typeof option.value !== null &&
+  typeof option.value !== 'undefined' &&
+  option.value === value;
+
+const Listbox = forwardRef<HTMLElement, ListboxProps>(
+  (
+    {
+      as: Component = 'ul',
+      children,
+      defaultValue,
+      value,
+      navigation = 'bound',
+      onKeyDown,
+      onFocus,
+      onSelect,
+      onSelectionChange,
+      onActiveChange,
+      ...props
+    },
+    ref
+  ): JSX.Element => {
+    const [options, setOptions] = useState<ListboxOption[]>([]);
+    const [activeOption, setActiveOption] = useState<ListboxOption | null>(
+      null
+    );
+    const [selectedOption, setSelectedOption] = useState<ListboxOption | null>(
+      null
+    );
+    const listboxRef = useSharedRef<HTMLElement>(ref);
+    const isControlled = typeof value !== 'undefined';
+
+    useLayoutEffect(() => {
+      const listboxValue = isControlled ? value : defaultValue;
+      const selectedOption = options.find((option) =>
+        optionMatchesValue(option, listboxValue)
+      );
+      setSelectedOption(selectedOption || null);
+      setActiveOption(selectedOption || null);
+    }, [options, value]);
```

Comment:
It seems like a bug would happen here if:
1. The component was uncontrolled
2. The options changed for some reason

We would reset the selected value to the default value.

Reply:
This PR isn't ready, you're probably looking for #1167